File: src/data/tutorials/004-trapping-rain-water.js
/**
 * Tutorial content for #004 — Trapping Rain Water
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given an array representing the height of bars in an elevation map, compute how much water can be trapped between the bars after it rains. Water sits on top of bars and is bounded on both sides by taller bars.`,
    example: `heights = [0,1,0,2,1,0,1,3,2,1,2,1]\n→ Water fills above index 2 (1 unit), above index 4 (1 unit), etc.\n✅ Answer: 6`,
    keyInsight: `The water level at any position is determined by the shorter of the tallest bar to its left and the tallest bar to its right. Water trapped at i = min(maxLeft[i], maxRight[i]) - height[i].`,
  },

  approaches: {
    'Brute Force': {
      intuition: `For every bar, scan left to find the tallest bar on the left, and scan right to find the tallest bar on the right. The water sitting on top of this bar is min(leftMax, rightMax) - height[i]. Sum this up for every position. Simple but does redundant scanning.`,
      steps: [
        `Initialize totalWater = 0.`,
        `For each index i from 0 to n-1:`,
        `  Scan left from i to 0, track leftMax = max height seen.`,
        `  Scan right from i to n-1, track rightMax = max height seen.`,
        `  water at i = min(leftMax, rightMax) - height[i].`,
        `  Add max(0, water) to totalWater.`,
        `Return totalWater.`,
      ],
      example: `heights = [0,1,0,2,1,0,1,3,2,1,2,1]\n\ni=2(h=0): leftMax=max(0,1,0)=1, rightMax=max(0,2,1,0,1,3,2,1,2,1)=3\n  water = min(1,3)-0 = 1\ni=5(h=0): leftMax=2, rightMax=3 → water=min(2,3)-0=2\n...\n✅ Total = 6`,
      keyInsight: `O(n²) time, O(1) space. For each of n positions we scan up to n positions — redundant work that the next approaches eliminate.`,
    },

    'Prefix & Suffix Arrays': {
      intuition: `Pre-compute the answers to "what is the tallest bar to my left?" and "what is the tallest bar to my right?" for every position. Store them in two arrays. Then a single pass computes the water at each position in O(1) using those pre-computed values.`,
      steps: [
        `Build leftMax[]: leftMax[i] = max of heights[0..i].`,
        `Build rightMax[]: rightMax[i] = max of heights[i..n-1].`,
        `For each index i: water[i] = min(leftMax[i], rightMax[i]) - heights[i].`,
        `Sum all water[i] values (negative values clamped to 0).`,
        `Return the total.`,
      ],
      example: `heights =  [0,1,0,2,1,0,1,3,2,1,2,1]\nleftMax  = [0,1,1,2,2,2,2,3,3,3,3,3]\nrightMax = [3,3,3,3,3,3,3,3,2,2,2,1]\n\ni=0: min(0,3)-0=0\ni=2: min(1,3)-0=1  ✓\ni=5: min(2,3)-0=2  ✓\ni=9: min(3,2)-1=1  ✓\n... total = 0+0+1+0+1+2+1+0+0+1+0+0 = 6\n✅ Answer: 6`,
      keyInsight: `O(n) time, O(n) space (two extra arrays). Clean and easy to understand — a great step up from brute force that illustrates the power of pre-computation.`,
    },

    'Two Pointers': {
      intuition: `Use two pointers starting at each end. The key insight: if the left bar is shorter, the water at the left pointer is fully determined by leftMax (the right side is guaranteed to have something at least as tall). Process the shorter side and move that pointer inward. This avoids needing any extra arrays.`,
      steps: [
        `Set left=0, right=n-1, leftMax=0, rightMax=0, totalWater=0.`,
        `While left < right:`,
        `  If heights[left] <= heights[right]: process the left side.`,
        `    Update leftMax = max(leftMax, heights[left]).`,
        `    Add leftMax - heights[left] to totalWater.`,
        `    left++.`,
        `  Else: process the right side symmetrically, right--.`,
        `Return totalWater.`,
      ],
      example: `heights = [0,1,0,2,1,0,1,3,2,1,2,1]\nleft=0,right=11,leftMax=0,rightMax=0\n\nh[0]=0≤h[11]=1: leftMax=0, water+=0-0=0, left=1\nh[1]=1>h[11]=1→no: rightMax=1, water+=1-1=0, right=10\nh[1]=1≤h[10]=2: leftMax=1, water+=1-1=0, left=2\nh[2]=0≤h[10]=2: leftMax=1, water+=1-0=1, left=3\n... (continuing)\n✅ Total = 6`,
      keyInsight: `O(n) time, O(1) space — the optimal solution. The two-pointer approach works because whichever side is shorter, its water level is already fully determined by the running max on that side.`,
    },

    'Monotonic Stack': {
      intuition: `Think of water filling horizontally between bars. Use a stack to keep track of bars that could form the left wall of a "valley". When you find a bar taller than the stack top, you've found the right wall — compute the water that fills the valley between the left wall and right wall, then pop and continue.`,
      steps: [
        `Initialize an empty stack and totalWater = 0.`,
        `Iterate over each index i:`,
        `  While stack is not empty AND heights[i] > heights[stack.top()]:`,
        `    Pop the top → this is the bottom of the valley.`,
        `    If stack is empty, break (no left wall).`,
        `    left = stack.top(), right = i.`,
        `    width = right - left - 1.`,
        `    boundedHeight = min(heights[left], heights[right]) - heights[bottom].`,
        `    totalWater += width * boundedHeight.`,
        `  Push i onto the stack.`,
        `Return totalWater.`,
      ],
      example: `heights = [0,1,0,2,...]\n\ni=0(0): stack=[0]\ni=1(1): h[1]>h[0]: pop 0 (bottom), stack empty→break. Push 1. stack=[1]\ni=2(0): h[2]<h[1]: push. stack=[1,2]\ni=3(2): h[3]>h[2]: pop 2(bottom=0), left=1(h=1),right=3(h=2)\n  width=1, height=min(1,2)-0=1, water+=1. stack=[1]\n  h[3]>h[1]: pop 1(bottom), stack empty→break. Push 3.\n... continue\n✅ Total = 6`,
      keyInsight: `O(n) time, O(n) space (stack). Computes water horizontally in layers — elegant for understanding the structure of trapped water. Each bar is pushed and popped at most once.`,
    },
  },
}

File: public/cpp/004-trapping-rain-water.cpp
/*
 * LeetCode Problem #42: Trapping Rain Water
 * Difficulty: Hard
 * Link: https://leetcode.com/problems/trapping-rain-water/
 * 
 * Problem Statement:
 * Given n non-negative integers representing an elevation map where the width
 * of each bar is 1, compute how much water it can trap after raining.
 */

#include <vector>
#include <stack>
#include <algorithm>
#include <iostream>
using namespace std;

class Solution {
public:
    // ==================== APPROACH 1: Brute Force ====================
    /*
     * Algorithm:
     * - For each position, find max height on left and right
     * - Water trapped at position = min(leftMax, rightMax) - height[i]
     * - Sum up water for all positions
     * 
     * Time Complexity: O(n²) - for each element, scan left and right
     * Space Complexity: O(1)
     * 
     * When to use: Only for understanding
     */
    int trap_BruteForce(vector<int>& height) {
        int n = height.size();
        int totalWater = 0;
        
        for (int i = 0; i < n; i++) {
            // Find max height on left
            int leftMax = 0;
            for (int j = 0; j <= i; j++) {
                leftMax = max(leftMax, height[j]);
            }
            
            // Find max height on right
            int rightMax = 0;
            for (int j = i; j < n; j++) {
                rightMax = max(rightMax, height[j]);
            }
            
            // Water at current position
            totalWater += min(leftMax, rightMax) - height[i];
        }
        
        return totalWater;
    }
    
    // ==================== APPROACH 2: Dynamic Programming ====================
    /*
     * Algorithm:
     * - Precompute leftMax and rightMax arrays
     * - leftMax[i] = max height from 0 to i
     * - rightMax[i] = max height from i to n-1
     * - Water at i = min(leftMax[i], rightMax[i]) - height[i]
     * 
     * Time Complexity: O(n) - three passes
     * Space Complexity: O(n) - two arrays
     * 
     * When to use: When you need O(n) time and can use O(n) space
     */
    int trap_DP(vector<int>& height) {
        int n = height.size();
        if (n == 0) return 0;
        
        vector<int> leftMax(n), rightMax(n);
        
        // Build leftMax array
        leftMax[0] = height[0];
        for (int i = 1; i < n; i++) {
            leftMax[i] = max(leftMax[i - 1], height[i]);
        }
        
        // Build rightMax array
        rightMax[n - 1] = height[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            rightMax[i] = max(rightMax[i + 1], height[i]);
        }
        
        // Calculate trapped water
        int totalWater = 0;
        for (int i = 0; i < n; i++) {
            totalWater += min(leftMax[i], rightMax[i]) - height[i];
        }
        
        return totalWater;
    }
    
    // ==================== APPROACH 3: Two Pointer (Optimal) ====================
    /*
     * Algorithm:
     * - Use two pointers from both ends
     * - Track leftMax and rightMax as we go
     * - Move pointer with smaller max height
     * - Add water based on the smaller max
     * 
     * Time Complexity: O(n) - single pass
     * Space Complexity: O(1) - only variables
     * 
     * When to use: This is the OPTIMAL solution
     * 
     * Key Insight:
     * - Water level at position depends on min(leftMax, rightMax)
     * - If leftMax < rightMax, we know water level is determined by leftMax
     * - We can calculate water at left position without knowing exact rightMax
     * - Same logic applies when rightMax < leftMax
     */
    int trap_TwoPointer(vector<int>& height) {
        int n = height.size();
        if (n == 0) return 0;
        
        int left = 0, right = n - 1;
        int leftMax = 0, rightMax = 0;
        int totalWater = 0;
        
        while (left < right) {
            if (height[left] < height[right]) {
                // Process left side
                if (height[left] >= leftMax) {
                    leftMax = height[left];
                } else {
                    totalWater += leftMax - height[left];
                }
                left++;
            } else {
                // Process right side
                if (height[right] >= rightMax) {
                    rightMax = height[right];
                } else {
                    totalWater += rightMax - height[right];
                }
                right--;
            }
        }
        
        return totalWater;
    }
    
    // ==================== APPROACH 4: Stack ====================
    /*
     * Algorithm:
     * - Use stack to store indices of bars
     * - When we find a bar taller than stack top, we can trap water
     * - Calculate water in horizontal layers
     * 
     * Time Complexity: O(n) - each bar pushed/popped once
     * Space Complexity: O(n) - stack storage
     * 
     * When to use: Alternative approach, useful for understanding
     * 
     * Key Insight:
     * - Stack maintains decreasing heights
     * - When we find taller bar, we can calculate trapped water
     * - Water is calculated layer by layer (horizontal slices)
     */
    int trap_Stack(vector<int>& height) {
        int n = height.size();
        stack<int> st;
        int totalWater = 0;
        
        for (int i = 0; i < n; i++) {
            // While current bar is taller than stack top
            while (!st.empty() && height[i] > height[st.top()]) {
                int top = st.top();
                st.pop();
                
                if (st.empty()) break;
                
                // Calculate water trapped
                int distance = i - st.top() - 1;
                int boundedHeight = min(height[i], height[st.top()]) - height[top];
                totalWater += distance * boundedHeight;
            }
            st.push(i);
        }
        
        return totalWater;
    }
};

// ==================== TEST CASES ====================
void runTests() {
    Solution sol;
    
    // Test Case 1: Standard case
    vector<int> height1 = {0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1};
    cout << "Test 1: " << sol.trap_TwoPointer(height1) << endl;
    // Expected: 6
    
    // Test Case 2: No water trapped
    vector<int> height2 = {3, 2, 1, 0};
    cout << "Test 2: " << sol.trap_TwoPointer(height2) << endl;
    // Expected: 0
    
    // Test Case 3: Simple valley
    vector<int> height3 = {3, 0, 2};
    cout << "Test 3: " << sol.trap_TwoPointer(height3) << endl;
    // Expected: 2
    
    // Test Case 4: Multiple valleys
    vector<int> height4 = {4, 2, 0, 3, 2, 5};
    cout << "Test 4: " << sol.trap_TwoPointer(height4) << endl;
    // Expected: 9
}

int main() {
    runTests();
    return 0;
}

/*
 * COMPARISON OF APPROACHES:
 * 
 * 1. Brute Force:
 *    Time: O(n²), Space: O(1)
 *    Not acceptable for interviews
 * 
 * 2. Dynamic Programming:
 *    Time: O(n), Space: O(n)
 *    Good solution, easy to understand
 * 
 * 3. Two Pointer (RECOMMENDED):
 *    Time: O(n), Space: O(1)
 *    Optimal solution, most elegant
 * 
 * 4. Stack:
 *    Time: O(n), Space: O(n)
 *    Alternative approach, calculates horizontally
 * 
 * INTERVIEW TIPS:
 * - Draw the elevation map to visualize
 * - Explain that water level = min(leftMax, rightMax)
 * - Start with DP approach, then optimize to two pointer
 * - Mention stack approach as alternative
 * - Discuss trade-offs between approaches
 * 
 * KEY INSIGHTS:
 * - Water at position depends on boundaries on both sides
 * - We only need the minimum of left and right max
 * - Two pointer exploits this by processing from smaller side
 * - Stack approach calculates water layer by layer
 * 
 * VISUALIZATION for [0,1,0,2,1,0,1,3,2,1,2,1]:
 * 
 *       █
 *   █xxx█x█
 * █x█x█████x█
 * 0102101321 21
 * 
 * Water (x) = 6 units
 */

// Made with Bob
